File: javascript/lang/networking/bind-to-localhost.js
const express = require('express');
const Fastify = require('fastify');

const app = express();
const port = Number(process.env.PORT ?? 8080);

// ruleid: bind-to-localhost
app.listen(port, '127.0.0.1');

const legacy = express();
// ruleid: bind-to-localhost
legacy.listen(3000, 'localhost', () => console.log('up'));

const fastify = Fastify();
// ruleid: bind-to-localhost
fastify.listen({ port, host: '127.0.0.1' });

const good = express();
// ok: bind-to-localhost
good.listen(port, '0.0.0.0');

const goodFastify = Fastify();
// ok: bind-to-localhost
goodFastify.listen({ port, host: '0.0.0.0' });

module.exports = { app, legacy, good };

File: python/lang/networking/bind-to-localhost.py
import os
import socket
from http.server import HTTPServer, SimpleHTTPRequestHandler

import uvicorn
from flask import Flask

app = Flask(__name__)


def serve_flask():
    # ruleid: bind-to-localhost
    app.run(host="127.0.0.1", port=8080)


def serve_asgi():
    # ruleid: bind-to-localhost
    uvicorn.run("main:app", host="localhost", port=8000)


def serve_http():
    # ruleid: bind-to-localhost
    HTTPServer(("127.0.0.1", 8000), SimpleHTTPRequestHandler).serve_forever()


def raw_socket():
    sock = socket.socket()
    # ruleid: bind-to-localhost
    sock.bind(("localhost", 9000))
    return sock


def serve_flask_all_interfaces():
    # ok: bind-to-localhost
    app.run(host="0.0.0.0", port=int(os.environ.get("PORT", "8080")))


def serve_asgi_all_interfaces():
    # ok: bind-to-localhost
    uvicorn.run("main:app", host="0.0.0.0", port=int(os.environ["PORT"]))
